include/pybind11/pytypes.h

```diff
@@ -798,11 +798,16 @@ PYBIND11_NAMESPACE_END(detail)
     : Parent(check_(o) ? o.release().ptr() : ConvertFun(o.ptr()), stolen_t{}) \
     { if (!m_ptr) throw error_already_set(); }
 
+#define PYBIND11_OBJECT_CHECK_FAILED(Name, o) \
+    type_error("Object of type '" + std::string(Py_TYPE(o.ptr())->tp_name) + "' mismatches '" #Name "'")
```

Comment:
How about `Object of type whatever is not an instance of Name`?

Reply:
Works. At least better than mine!